### PR TITLE
[WIP] fixes restyling for stem plots

### PR DESCRIPTION
**Pull request by ssfrr**
Seems to be working, and passes existing tests. I also added new tests for the previous stem plot behavior and now restyling. To test that the javascript stuff is getting called correctly I added a `MockPlotlyDisplay` type that partially mocks the API for the other displays so I can make sure things are getting called correctly.

There are a few cleanup things that I think would be good to do, but they involve more change to existing code, so I wanted to check in beforehand to see if you're on-board, and/or whether you'd want these changes as a separate PR:

* [ ] refactor `AbstractPlotlyDisplay` API so that the low-level methods (like `restyle!`) don't do any argument transformation and JUST forward the calls through to javascript. That way the `MockPlotlyDisplay` doesn't have to copy-paste behavior from the other displays.
* [ ] bump up `MockPlotlyDisplay` to top-level of tests so it can be used by other tests outside the `stem` stuff
* [ ] move additional `scatter` test in with the rest of the `scatter` tests
* [ ] the fact that many functions (internal and external) use both an update dict and keyword args complicates a lot of the implementation. I propose we merge/normalize to a dict as soon as we get the args from the user, and then all the internal functions just use the dict.

### CODE REVIEW

File: src/display.jl [hunk 1]
```diff
@@ -75,7 +75,22 @@ end
 plot(args...; kwargs...) = SyncPlot(Plot(args...; kwargs...))
 
 ## API methods for SyncPlot
-for f in [:restyle!, :relayout!, :addtraces!, :deletetraces!, :movetraces!,
+function restyle!(sp::SyncPlot, args...; kwargs...)
+    diffs = restyle!(sp.plot, args...; kwargs...)
+    # each trace in the plot returns a diff, so we have to apply them
+    # separately. Maybe at some point we can do some clever merging
+    for (i, diff) in enumerate(diffs)
```
Comment (spencerlyon2) [inline src/display.jl]: This is a good first pass, but I think that we'll eventually want to merge all the diffs on the Julia side so we aren't sending many little diffs over the server to js.

I could be wrong, but I anticipate that performance would be better if we did the merging in Julia

**Comment by ssfrr** [comment]
Something seems to have gone sideways on the OSX build. It's marked as failed but I don't see any output in the log.

File: src/display.jl [hunk 2]
```diff
@@ -75,7 +75,22 @@ end
...
+        restyle!(sp.view, i, diff)
+    end
+end
+
+function restyle(sp::SyncPlot, args...; kwargs...)
+    sp2 = fork(sp)
+    $f(sp2.plot, args...; kwargs...)  # only need to update the julia side
```
Comment (spencerlyon2) [inline src/display.jl]: Is the `$f` here a typo?

File: src/display.jl (hunk 1 above)
Comment (spencerlyon2) [inline src/display.jl]: Hmm, I'm looking closer at this and how can we be sure that the `i` from enumerate is the actual `i` the user chose when he was restyling a specific trace?

I have some ideas for how to do the merging of diffs on the Julia side, but want to make sure I understand everything you have right now first

**Pull request closed without merging**